k2/csrc/fsa.cc

```diff
@@ -0,0 +1,250 @@
+  size_t arcs_array_bytes = sizeof(Arc) * cfsa.NumArcs();
+  res_bytes += arcs_array_bytes;
+
+  return res_bytes;
+}
+
+size_t GetCfsaVecSize(const std::vector<Cfsa> &cfsas) {
+  size_t res_bytes = 0;
+
+  size_t header_bytes = sizeof(CfsaVecHeader);
+  res_bytes += header_bytes;
+
+  // padding to the alignment boundary for state_offsets_array
+  res_bytes = AlignTo(res_bytes, kAlignment);
+
+  // size in bytes for `int32_t state_offsets_array[num_fsas + 1];`
+  size_t state_offsets_array_bytes = sizeof(int32_t) * (cfsas.size() + 1);
+  res_bytes += state_offsets_array_bytes;
+
+  // padding to the alignment boundary for arc_indexes_array
+  res_bytes = AlignTo(res_bytes, kAlignment);
+
+  int32_t num_states = 0;
+  int32_t num_arcs = 0;
+  for (const auto &cfsa : cfsas) {
+    num_states += cfsa.NumStates();
+    num_arcs += cfsa.NumArcs();
+  }
+
+  // size in bytes for `int32_t arc_indexes_array[num_states + num_fsas];`
+  size_t arc_indexes_array_bytes =
+      sizeof(int32_t) * (num_states + cfsas.size());
+  res_bytes += arc_indexes_array_bytes;
+
+  // align res_bytes to be multiple of sizeof(Arc)
+  res_bytes = (res_bytes + sizeof(Arc) - 1) / sizeof(Arc) * sizeof(Arc);
+
+  DCHECK_EQ(res_bytes % alignof(Arc), 0);
+
+  // size in bytes for `Arc arcs[num_arcs];`
+  size_t arcs_array_bytes = sizeof(Arc) * num_arcs;
+  res_bytes += arcs_array_bytes;
+
+  return res_bytes;
+}
+
+void CreateCfsaVec(const std::vector<Cfsa> &cfsas, void *data, size_t size) {
+  DCHECK_EQ(size, GetCfsaVecSize(cfsas));
+
+  auto header = reinterpret_cast<CfsaVecHeader *>(data);
+  header->version = kCfsaVecVersion;
+  header->num_fsas = static_cast<int32_t>(cfsas.size());
+
+  size_t offset = sizeof(CfsaVecHeader);
+
+  offset = AlignTo(offset, kAlignment);
+  header->state_offsets_start = offset / sizeof(int32_t);
+
+  auto state_offsets_array =
+      reinterpret_cast<int32_t *>(data) + header->state_offsets_start;
+
+  int32_t num_states = 0;
+
+  state_offsets_array[0] = 0;
+  int32_t i = 1;
+  for (; i < static_cast<int32_t>(cfsas.size()); ++i) {
+    if (cfsas[i - 1].NumStates() != 0) {
+      state_offsets_array[i] =
+          state_offsets_array[i - 1] + cfsas[i - 1].NumStates() + 1;
+    } else {
+      state_offsets_array[i] = state_offsets_array[i - 1];
+    }
+
+    num_states += cfsas[i].NumStates();
+  }
+
+  if (!cfsas.empty()) {
+    if (cfsas.back().NumStates() != 0) {
+      state_offsets_array[i] =
+          state_offsets_array[i - 1] + cfsas.back().NumStates() + 1;
+    } else {
+      state_offsets_array[i] = state_offsets_array[i - 1];
+    }
+
+    num_states += cfsas[0].NumStates();
+  }
+
+  offset += sizeof(int32_t) * (cfsas.size() + 1);
+  offset = AlignTo(offset, kAlignment);
+  header->arc_indexes_start = offset / sizeof(int32_t);
+
+  auto arc_indexes_array =
+      reinterpret_cast<int32_t *>(data) + header->arc_indexes_start;
+  offset += sizeof(int32_t) * (num_states + cfsas.size());
```

Comment:
it would be greate if we can add comments as you do in `GetCfsaVecSize(const Cfsa &cfsa)` to clarify the purpose when updating `offset` each time, as well the `+csfsa.size()` trick (as we add a duplicate final state)

Reply:
fixed.